### PR TITLE
[Feat] Add Lab color space

### PR DESCRIPTION
**Pull request by cceyda**
### Description
Continuing from https://github.com/kornia/kornia/pull/350

Since  rgb<->xyz implementations were already there I just followed along the luv implemetation and added xyz<->lab stuff.

referencing:
https://github.com/torch/image/blob/dc061b98fb7e946e00034a5fc73e883a299edc7f/generic/image.c#L1467
https://github.com/scikit-image/scikit-image/blob/a48bf6774718c64dade4548153ae16065b595ca9/skimage/color/colorconv.py

https://docs.opencv.org/master/de/d25/imgproc_color_conversions.html
http://www.brucelindbloom.com/index.html?Eqn_Lab_to_XYZ.html
http://www.brucelindbloom.com/index.html?ChromAdaptEval.html

https://github.com/richzhang/colorization-pytorch/blob/master/util/util.py

~still in_progress because: https://github.com/xahidbuffon/TF_RGB_LAB/blob/master/rgb_lab_formulation.py might have a more efficient implementation.~ Edit: forgoing that for sake of current clearer implementation, wasn't much to gain.

### Status
**Work in progress**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/color/lab.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    gs: torch.Tensor = rgbs_im[..., 1, :, :]
+    bs: torch.Tensor = rgbs_im[..., 2, :, :]
+
+    r: torch.Tensor = torch.where(rs > 0.0031308, 1.055 * torch.pow(rs, 1 / 2.4) - 0.055, 12.92 * rs)
+    g: torch.Tensor = torch.where(gs > 0.0031308, 1.055 * torch.pow(gs, 1 / 2.4) - 0.055, 12.92 * gs)
+    b: torch.Tensor = torch.where(bs > 0.0031308, 1.055 * torch.pow(bs, 1 / 2.4) - 0.055, 12.92 * bs)
+
+    rgb_im: torch.Tensor = torch.stack([r, g, b], dim=-3)
+
+    # Clip to 0,1 https://www.w3.org/Graphics/Color/srgb
```
Comment (cceyda) [inline kornia/color/lab.py:138]: [scikit-image](https://github.com/scikit-image/scikit-image/blob/master/skimage/color/colorconv.py#L618) does clipping, but if we do clip we won't have forth_and_back -> `rgb != lab2rgb(rgb2lab(rgb))` ... what to do?

**Review by cceyda — COMMENTED** [review]
I finally made a contained env to work on this and tests passed there. Looks like all the tests fail here due to some conda thing, will wait until it is fixed.

File: kornia/color/lab.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+    fz = torch.where(fz < 0, torch.zeros_like(fz), fz)
+
+    fxyz = torch.stack([fx, fy, fz], dim=-3)
+
+    # Convert from Lab to XYZ
+    xyz: torch.Tensor = torch.where(fxyz > .2068966, torch.pow(fxyz, 3.0), (fxyz - 4. / 29.) / 7.787)
+
+    # For D65 white point
+    xyz_ref_white: torch.Tensor = torch.tensor([0.95047, 1., 1.08883])[..., :, None, None]
```
Comment (cceyda) [inline kornia/color/lab.py:121]: Should we also support different [illuminants](https://github.com/scikit-image/scikit-image/blob/2db874cc6485186b08d296887a8a91bb2f8865ec/skimage/color/colorconv.py#L419)? or put these values somewhere more central

Comment (edgarriba) [reply kornia/color/lab.py:121]: maybe as a dict on the top of this file for now. How would you define the api for different illuminants - `mode='D65'` and so on ? I suggest adding it in a new PR with proper tests.

File: kornia/color/lab.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    xyz_ref_white = xyz_ref_white.to(xyz.device)
```
Comment (edgarriba) [inline kornia/color/lab.py:122]: same as before for dtype

File: kornia/color/lab.py (hunk 1 above)
Comment (edgarriba) [reply kornia/color/lab.py:138]: how does the output affect at all w/o the clipping ?

**Review by edgarriba — CHANGES_REQUESTED** [review]

**Comment by edgarriba** [comment]
@cceyda don't forget to add to docs, otherwise it won't render to readthedocs

**Comment by edgarriba** [comment]
@cceyda not sure what's going on with tests. Could you try to rebase from master ?

File: kornia/color/lab.py (hunk 1 above)
Comment (cceyda) [reply kornia/color/lab.py:131]: I have tested what kind of values we get by using an image with all possible RGB values [0,255]. Opencv vs skimage vs kornia, all here in this gist
https://gist.github.com/cceyda/2a597b293782b07169328a92c7f7e979
The values that fall outside of [0,1] don't do so by much (~e-07). But since it effects the reversibility of the operation I think we can return the unclipped values (with a warning?) and leave it up to the user (like some other [libraries](https://python-colormath.readthedocs.io/en/latest/conversions.html#rgb-conversions-and-out-of-gamut-coordinates)) . OR do optional clipping with an clip=bool argument?

Other than that;
The current code produces the same results as skimage (without clipping).
There are some minor differences with opencv I'm assuming because of C vs python float op. differences. (Not sure if open cv does clipping)

File: kornia/color/lab.py (hunk 2 above)
Comment (cceyda) [reply kornia/color/lab.py]: Agree to leave it for another PR if there is demand for it.

File: kornia/color/lab.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    scale = 7.787 * xyz_normalized + 4. / 29.
+    xyz_int = torch.where(xyz_normalized > 0.008856, power, scale)
+
+    x: torch.Tensor = xyz_int[..., 0, :, :]
+    y: torch.Tensor = xyz_int[..., 1, :, :]
+    z: torch.Tensor = xyz_int[..., 2, :, :]
+
+    L: torch.Tensor = (116. * y) - 16.
+    a: torch.Tensor = 500. * (x - y)
+    b: torch.Tensor = 200. * (y - z)
```
Comment (edgarriba) [inline kornia/color/lab.py:68]: mypy complains about this variable since you are overwriting it. To avoid future debugging issues I think it could be renamed to `b_new`. Same for `lab_to_rgb`.

**Review by edgarriba — COMMENTED** [review]
@cceyda I think the rebase was not done corrctly. Try with `git pull --rebase upstream/master` and possibly force push later

File: kornia/color/lab.py (hunk 1 above)
Comment (edgarriba) [reply kornia/color/lab.py:131]: I wasn't aware of `python-colormath` - looks like a good image color mine :)
not sure raising a warning it's a good practice since the user using the functionality with the warning branch will receive annoying outputs every-time the function is called. Probably the `clip: bool` parameters is a better approach. /cc @ducha-aiki 
I believe other functionalities in this module can be adapted following this same convention.

**Comment by edgarriba** [comment]
@cceyda any update on this ? seems that the rebase didn't happen as expected. We are releasing v0.5 in a week - would be great to include this feature.

**Comment by cceyda** [comment]
Sorry I have been very busy~ but I will get to it over the weekend.

**Comment by cceyda** [comment]
I have added clipping option to `lab2rgb` [clip=True is the default]
Right now the calculations are the same as scikit-image. But If you want to really fall down the rabbit hole that is colorspaces 😵  (I don't think we should here) this library is a good resource: https://colour.readthedocs.io/en/latest/

- The TPU test are failing because margin of error is bigger than 0.0001🤷  It looks like a high 0.01 margin is happening. 

A couple of FYIs:
- The clipping is done as part of xyz2rgb conversion (same as scikit-image)
- For `lab2rgb` if lab values are not within normal ranges, include too small values etc (as might be the case if they are the outputs of a neural net) we might get possible nans. As warned against [here](https://github.com/richzhang/colorization-pytorch/blob/66a1cb2e5258f7c8f374f582acc8b1ef99c13c27/util/util.py#L107). But if I add that line lab-rgb forward-backward breaks. And I don't see a reason why the user can't ensure the input range to the function. The lab2rgb is usually the last part of the pipeline anyway. Mentioning it here because neural colorization is a common use case for lab<->rgb conversions (and part of my motivation for doing this)
 
Let's include this as a kind of beta and see. I wouldn't mind babysitting it for a while.
I might have committed a git crime by force pushing twice 🙈

**Comment by edgarriba** [comment]
@cceyda for tpu we are still investigating what's the reason of this drop for the accuracy. In the meantime, we came with this hack: https://github.com/kornia/kornia/blob/master/test/enhance/test_zca.py#L83-L84. Seems that xla devices are less precise in general.
As for the library you mention, do you think it's worth to adopt their standards in future ?

**Comment by edgarriba** [comment]
@cceyda seems that we have an issue with our gcp account - don't alarm with tpu tests :)

**Comment by cceyda** [comment]
> @cceyda for tpu we are still investigating what's the reason of this drop for the accuracy. In the meantime, we came with this hack: https://github.com/kornia/kornia/blob/master/test/enhance/test_zca.py#L83-L84. Seems that xla devices are less precise in general.

I have changed the test precision the same way as the linked code. (tests now fail due to authentication)

> As for the library you mention, do you think it's worth to adopt their standards in future ?
There is a lot of variance across libraries when it comes to color conversions (due to precision, input type and language differences). IMO, That is a very good in-depth library about color related operations (for referencing) it helped me understand some operations.  But I don't think we need to go as far as to implement similar functionality.

Right now in kornia we have a bit of a concept confusion regarding RGB, sRGB and linear rgb in the code variable naming etc (it is a common source of confusion). Next time, in another pr, when I have a bit of time I will refactor the sRGB<->linear RGB stuff out (so that `luv` and `lab` can just call that). Also maybe adjust the `xyz` conversion matrix precision.

**Comment by edgarriba** [comment]
@cceyda that would be awesome from your side.

**Review by edgarriba — APPROVED** [review]
